### PR TITLE
[MIR] Enhance the SimplifyCfg pass to merge consecutive blocks

### PR DESCRIPTION
**Pull request by dotdash**
Updated from #30238, including the changes suggested by @Aatch.

### CODE REVIEW

File: src/librustc_mir/transform/simplify_cfg.rs [hunk 1]
```diff
@@ -22,59 +25,136 @@ impl SimplifyCfg {
...
+}
+
+// Find the target at the end of the jump chain, return None if there is a loop
+fn final_target(mir: &Mir, mut target: BasicBlock) -> Option<BasicBlock> {
+    // Keep track of already seen blocks to detect loops
+    let mut seen: Vec<BasicBlock> = Vec::with_capacity(8);
+
+    while mir.basic_block_data(target).statements.is_empty() {
+        // NB -- terminator may have been swapped with `None`
+        // below, in which case we have a cycle and just want
```
Comment (luqmana) [inline src/librustc_mir/transform/simplify_cfg.rs]: Nit: I think the `below` here became above. Could just say in the function `merge_consecutive_blocks` instead.

**Comment by luqmana** [comment]
LGTM! 👍

File: src/librustc_mir/transform/predecessor_map.rs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    }
+}
+
+impl<'tcx> Visitor<'tcx> for PredecessorVisitor {
+    fn visit_branch(&mut self, source: BasicBlock, target: BasicBlock) {
+        self.predecessor_map.add_predecessor(target, source);
+    }
+}
+
+pub fn build_predecessor_map(mir: &Mir) -> PredecessorMap {
```
Comment (nagisa) [inline src/librustc_mir/transform/predecessor_map.rs]: NIT: to me it seems like it ought to be a `PredecessorMap::from_mir` or something.

File: src/librustc_mir/transform/simplify_cfg.rs [hunk 3]
```diff
@@ -22,59 +25,136 @@ impl SimplifyCfg {
...
+
+                            predecessor_map.replace_predecessor(target, bb, target);
+                            for succ in terminator.successors().iter() {
+                                predecessor_map.replace_predecessor(*succ, target, bb);
+                            }
+                        }
+                        TerminatorKind::Goto { target: new_target } if num_insts == 0 => {
+                            inner_change = true;
+                            terminator.kind = TerminatorKind::Goto { target: new_target };
+                            predecessor_map.remove_predecessor(target, bb);
```
Comment (nagisa) [inline src/librustc_mir/transform/simplify_cfg.rs]: replace_predecessor.

Comment (dotdash) [inline src/librustc_mir/transform/simplify_cfg.rs]: Good idea!

**Comment by dotdash** [comment]
Rebased and nits addressed.

@bors r=luqman

**Pull request merged**